### PR TITLE
Handle Variables in the URL

### LINKED ISSUE #45: Multi Page Apps [feature request]: variables in the path
Add the ability to have variables in the path - as suggested by [@benn0 on the forum](https://community.plotly.com/t/introducing-dash-pages-a-dash-2-1-feature-preview/57775/4?u=annmariew)

### PR DESCRIPTION
**Pull request by AnnMarieW**
Closes #45 

This is a new feature for handing variables in the URL.

Creates a new parameter `path_template` in `dash.page_registry` 

- `path_template`:
       Add variables to a URL by marking sections with <variable_name>. The layout function
       then receives the <variable_name> as a keyword argument.
       e.g. path_template= "/asset/<asset_id>"
             path = "/assets/a100" 
            layout will receive `{"asset_id":"a100"}`

If no `path` is supplied when calling `dash.register_page()` then it will be inferred from the `path_template`.  Variables will default to `"none"`.
  e.g. "/asset/<asset_id>"  => "/asset/none"

### CODE REVIEW

File: dash_labs/plugins/pages.py [hunk 1]
```diff
@@ -259,12 +277,20 @@ def router():
...
                 page = dash.page_registry[module]
+
                 if path_id == app.strip_relative_path(page["path"]):
                     layout = page["layout"]
 
+                if page["path_template"]:
+                    template_id = app.strip_relative_path(page["path_template"])
+                    path_variables = _parse_path_variables(path_id, template_id)
+                    if path_variables:
+                        layout = page["layout"]
```
Comment (alexcjohnson) [inline dash_labs/plugins/pages.py:292]: 🐎 do we want to `break` here and [above](https://github.com/plotly/dash-labs/pull/61/files#diff-3242435debccc1e6c9e9164d3826cd066d02302d12cda2cb20956e71d27da55eR286) since we found a matching page?

Comment (AnnMarieW) [reply dash_labs/plugins/pages.py:292]: For sure - that would be more efficient

File: dash_labs/plugins/pages.py [hunk 2]
```diff
@@ -398,3 +430,24 @@ def _parse_query_string(search):
...
+        return None
+
+    path_vars = {}
+    for i in range(len(path_segments)):
+        if path_segments[i] == template_segments[i]:
+            continue
+        if template_segments[i].startswith("<"):
+            # removes the "<" ">" around the variable key
+            variable_key = template_segments[i][1:-1]
+            path_segments[i] = None if path_segments[i] == "None" else path_segments[i]
```
Comment (alexcjohnson) [inline dash_labs/plugins/pages.py:449]: This `"None"` -> `None` conversion worries me - what if you really did want to pass `"None"` as the parameter? Can we just leave it as `"None"`? I know up [above](https://github.com/plotly/dash-labs/pull/61/files#diff-3242435debccc1e6c9e9164d3826cd066d02302d12cda2cb20956e71d27da55eR247) we use `"None"` in the path when registering the page but I'm not sure this is a good idea either - for example, how is that going to work in the title, description, and image?

https://github.com/plotly/dash-labs/blob/f182bc011291fedce47c13e6a889d09b17a7c020/dash_labs/plugins/pages.py#L324-L332

I suspect the better way to handle this will be to take the [path-to-layout finder](https://github.com/plotly/dash-labs/pull/61/files#diff-3242435debccc1e6c9e9164d3826cd066d02302d12cda2cb20956e71d27da55eR282-R292) and turn it into a path-to-page finder, then use that both here and in `interpolate_index`

File: dash_labs/plugins/pages.py [hunk 3]
```diff
@@ -398,3 +430,24 @@ def _parse_query_string(search):
...
+
+def _parse_path_variables(pathname, path_template):
+    path_segments = pathname.split("/")
+    template_segments = path_template.split("/")
+
+    if len(path_segments) != len(template_segments):
+        return None
+
+    path_vars = {}
+    for i in range(len(path_segments)):
```
Comment (alexcjohnson) [inline dash_labs/plugins/pages.py:443]: this would be a little cleaner as:
`for path_segment, template_segment in zip(path_segments, template_segments):`
since you already know the lengths are the same.

That said I suspect at some point we'll want to extend this to support [Flask converter types](https://flask.palletsprojects.com/en/2.0.x/quickstart/#variable-rules) including `path` that also allows slashes in the match, but that will need some special logic anyway.

Comment (AnnMarieW) [reply dash_labs/plugins/pages.py:443]: I thought about doing the zip, but then there would be additional logic to pull out the variables

File: dash_labs/plugins/pages.py (hunk 2 above)
Comment (alexcjohnson) [reply dash_labs/plugins/pages.py:449]: In fact, I could imagine users wanting title, description, and image to each potentially be a function of the path variables... But for now we should just ensure they work correctly as constant values for each registered page.

File: dash_labs/plugins/pages.py (hunk 3 above)
Comment (alexcjohnson) [reply dash_labs/plugins/pages.py:443]: Where? All I see is `path_segments[i]` which would just turn into `path_segment`, and similarly with `template_segments[i]`->`template_segment`. You never use `i` for any purpose other than pulling out the ith element of those two lists, so you should be able to get rid of it entirely.

File: dash_labs/plugins/pages.py (hunk 1 above)
Comment (AnnMarieW) [reply dash_labs/plugins/pages.py:294]: actually, I could add a `break` here but not [above](https://github.com/plotly/dash-labs/pull/61/files#diff-3242435debccc1e6c9e9164d3826cd066d02302d12cda2cb20956e71d27da55eR286) because above can be true for paths with templates

File: dash_labs/plugins/pages.py (hunk 3 above)
Comment (AnnMarieW) [reply dash_labs/plugins/pages.py]: I have a new version [here](https://github.com/plotly/dash-labs/pull/61/commits/9c5e92bc6cd5c9c4437b1a64face96ea085bc5a2) is it better?

I also removed the conversion from "None" to `None`.   

Also, When no `path` is supplied to  `dash.page_registry` I changed the path variables in the inferred path  "none"  (instead of "None").  I hope this makes it more clear that it's a string and not actually `None`

File: dash_labs/plugins/pages.py (hunk 2 above)
Comment (AnnMarieW) [reply dash_labs/plugins/pages.py]: Actually this works without any changes.  I updated the example app to include `title` and `description`.  Here's what the meta tags look like when no initial `path` is provided:
 
image

And here is what it looks like when a different (valid) initial path is provided:
image

Comment (AnnMarieW) [reply dash_labs/plugins/pages.py]: > In fact, I could imagine users wanting title, description, and image to each potentially be a function of the path variables... But for now we should just ensure they work correctly as constant values for each registered page.

I think one way to do this is to create a `dash.path_registry`.  It could start by being generated from `dash.page_registry`  Key would be the `path`, values would include `title, description, image`.  Then people could add other paths.

**Comment by AnnMarieW** [comment]
Hey @alexcjohnson This is ready for another review.  I moved tests to #66 so it doesn't hold up things that are finished, such as:


- Fixed the "None" => `None` conversion.
- Confirmed that when there are variables in the path, the meta tags for title, description, and image are correct when the page first loads - both when the path is supplied or inferred.  However it doesn't update when the path variables change.  If that's a popular feature, then I think it could be added later.
- The `assets/` and `pages/` folders are now optional.  If  `pages/` can't be found,  it no longer raises  an Exception,  but it does print a message.  The message  could help debug if not finding  `pages/` is unintentional.


- There is now a check that the paths are unique for each module in dash.page_registry. This should help prevent bugs.
- There is now a check that pathnames start with a `/`.   Previously the `/` was missing in some cases when the layout was added directly to dash.register_page()

- Updated docs and included new info on path variables and query strings

File: dash_labs/plugins/pages.py (hunk 2 above)
Comment (AnnMarieW) [reply dash_labs/plugins/pages.py]: >I suspect the better way to handle this will be to take the path-to-layout finder and turn it into a path-to-page finder, then use that both here and in interpolate_index.

Good suggestion - I added a `path_to_page` function in  [this commit](https://github.com/plotly/dash-labs/pull/61/commits/2ddad2ca6f89c45e35fc05fd90139d5cd4276c96)

File: docs/08-MultiPageDashApp.md [hunk 4]
```diff
@@ -285,47 +323,63 @@ page navigation links or by template authors.
...
    The module path where this page's `layout` is defined. Often `__name__`.
 
+
 - `path`:
    URL Path, e.g. `/` or `/home-page`.
    If not supplied, will be inferred from `module`,
    e.g. `pages.weekly_analytics` to `/weekly-analytics`
 
+
+- path_template:
```
Comment (chriddyp) [inline docs/08-MultiPageDashApp.md:333]:
```suggestion
- `path_template`:
```

File: dash_labs/plugins/pages.py [hunk 5]
```diff
@@ -9,22 +9,28 @@
...
+import re
 
-if not os.path.exists("pages"):
-    raise Exception("A folder called `pages` does not exist.")
 
 _ID_CONTENT = "_pages_plugin_content"
 _ID_LOCATION = "_pages_plugin_location"
-_ID_DUMMY = "_pages_plugin_dummy"
+_ID_STORE = "_pages_plugin_store"
+_ID_DUMMY = "pages_plugin_dummy"
```
Comment (alexcjohnson) [inline dash_labs/plugins/pages.py:18]:
```suggestion
_ID_DUMMY = "_pages_plugin_dummy"
```
Was that change intentional?

Comment (AnnMarieW) [reply dash_labs/plugins/pages.py:18]: oops that was a typo

File: dash_labs/plugins/pages.py [hunk 6]
```diff
@@ -224,16 +241,17 @@ def _filename_to_name(filename):
...
-    return filename.replace("_", "-").replace(".", "/").lower().split("pages")[-1]
-
+def _infer_path(filename, template):
+    if template is None:
+        path = filename.replace("_", "-").replace(".", "/").lower().split("pages")[-1]
+        path = "/" + path if not path.startswith("/") else path
+        return path
+    else:
+        # replace the variables in the template with "None"
+        return re.sub("<.+?>", "none", template)
```
Comment (alexcjohnson) [inline dash_labs/plugins/pages.py:251]:
```suggestion
        return re.sub("<[^<>]+>", "none", template)
```
Not a big deal, I'm just brainstorming ways the user could break this, and at present unmatched or empty angle brackets would produce very weird results.

Maybe in fact the best way to handle this would be validating the template in `register_page`: break it into segments like we do in `_parse_path_variables`, but then for any segment that has either `<` or `>` in the segment, throw an error if the segment doesn't match `^<[^<>]+>$`

Comment (AnnMarieW) [reply dash_labs/plugins/pages.py:251]: ah, yes - good idea. I'll add that

**Review by alexcjohnson — APPROVED** [review]
Looks good! Stricter pattern checking https://github.com/plotly/dash-labs/pull/61/files#r778960479 would be nice but we could do that later as well.
💃

File: dash_labs/plugins/pages.py [hunk 7]
```diff
@@ -224,16 +243,27 @@ def _filename_to_name(filename):
     return filename.split(".")[-1].replace("_", " ").capitalize()
 
 
-def _filename_to_path(filename):
-    return filename.replace("_", "-").replace(".", "/").lower().split("pages")[-1]
+def _validate_template(template):
+    template_segments = template.split("/")
+    for s in template_segments:
+        if ("<" or ">") in s and not (s.startswith("<") and s.endswith(">")):
```
Comment (alexcjohnson) [inline dash_labs/plugins/pages.py:249]: `("<" or ">") in s`  -> I think you mean `("<" in s or ">" in s)` 

Also need to check that `<` and `>` are NOT in `s[1: -1]`

Comment (AnnMarieW) [reply dash_labs/plugins/pages.py]: > Also need to check that < and > are NOT in s[1: -1]

Just to summarize our discussion:  Since the variables are passed to the layout in a dict as strings, there should not be a restriction about including  `<` or ` >` .   However, the most common use case is that the keys are Python variables, so now there is a warning message if the variable is not a valid Python variable name.

**Review by alexcjohnson — APPROVED** [review]
Alright I think we have a winner! 💃
